File: skills/supermap-wiki/scripts/search_wiki.py
#!/usr/bin/env python3
"""
Supermap Wiki Search Script

Searches the Supermap Confluence wiki and returns results as a markdown table.
Uses only Python standard library for cross-platform compatibility.
"""

import argparse
import json
import os
import sys
import urllib.error
import urllib.parse
import urllib.request


def get_token():
    """Get the wiki token from environment variable."""
    token = os.environ.get('SUPERMAP_WIKI_TOKEN')
    if not token:
        print("Error: SUPERMAP_WIKI_TOKEN environment variable is not set.", file=sys.stderr)
        print("Please set it using:", file=sys.stderr)
        print("  Linux/macOS: export SUPERMAP_WIKI_TOKEN='your-token'", file=sys.stderr)
        print("  Windows (cmd): set SUPERMAP_WIKI_TOKEN=your-token", file=sys.stderr)
        print("  Windows (PowerShell): $env:SUPERMAP_WIKI_TOKEN='your-token'", file=sys.stderr)
        sys.exit(1)
    return token


def build_search_url(query, start=0, limit=20):
    """Build the search API URL with query parameters."""
    base_url = "https://wiki.ispeco.com/rest/api/search"

    # Build CQL query
    cql = f'siteSearch ~ "{query}" AND type in ("space","user","com.atlassian.confluence.extra.team-calendars:calendar-content-type","attachment","page","com.atlassian.confluence.extra.team-calendars:space-calendars-view-content-type","blogpost")'

    params = {
        'cql': cql,
        'start': start,
        'limit': limit,
        'excerpt': 'highlight',
        'expand': 'space.icon',
        'includeArchivedSpaces': 'false'
    }

    # URL encode parameters
    encoded_params = urllib.parse.urlencode(params)
    return f"{base_url}?{encoded_params}"


def search_wiki(query, token, limit=20):
    """Perform the wiki search and return results."""
    url = build_search_url(query, limit=limit)

    # Create request with headers
    req = urllib.request.Request(url)
    req.add_header('Authorization', f'Bearer {token}')
    req.add_header('Accept', 'application/json')

    try:
        with urllib.request.urlopen(req, timeout=30) as response:
            data = json.loads(response.read().decode('utf-8'))
            return data
    except urllib.error.HTTPError as e:
        if e.code == 401:
            print("Error: Authentication failed. Please check your SUPERMAP_WIKI_TOKEN.", file=sys.stderr)
        elif e.code == 403:
            print("Error: Access forbidden. You may not have permission to search the wiki.", file=sys.stderr)
        else:
            print(f"Error: HTTP {e.code} - {e.reason}", file=sys.stderr)
        sys.exit(1)
    except urllib.error.URLError as e:
        print(f"Error: Network error - {e.reason}", file=sys.stderr)
        print("Please check your network connection.", file=sys.stderr)
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Failed to parse response - {e}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error: Unexpected error - {e}", file=sys.stderr)
        sys.exit(1)


def format_as_markdown(results):
    """Format search results as a markdown table."""
    if not results:
        print("No results found.")
        return

    # Table header
    print("| Title | Space | Excerpt |")
    print("|-------|-------|---------|")

    # Table rows
    for result in results:
        content = result.get('content', {})
        title = content.get('title', 'N/A')

        # Get space name
        container = result.get('resultGlobalContainer', {})
        space = container.get('title', 'N/A')

        # Get excerpt and clean it up
        excerpt = result.get('excerpt', '')
        # Remove HTML tags and clean whitespace
        excerpt = excerpt.replace('<span class="search-highlight">', '**')
        excerpt = excerpt.replace('</span>', '**')
        excerpt = excerpt.replace('\n', ' ').strip()
        # Truncate long excerpts
        if len(excerpt) > 150:
            excerpt = excerpt[:147] + '...'

        # Build link
        webui_link = content.get('_links', {}).get('webui', '')
        if webui_link:
            link = f"https://wiki.ispeco.com{webui_link}"
            title = f"[{title}]({link})"

        # Escape pipe characters in content
        title = title.replace('|', '\\|')
        space = space.replace('|', '\\|')
        excerpt = excerpt.replace('|', '\\|')

        print(f"| {title} | {space} | {excerpt} |")


def main():
    parser = argparse.ArgumentParser(
        description='Search Supermap wiki and display results as a markdown table.'
    )
    parser.add_argument(
        'query',
        help='Search query string'
    )
    parser.add_argument(
        '-l', '--limit',
        type=int,
        default=20,
        help='Maximum number of results (default: 20)'
    )

    args = parser.parse_args()

    # Get token
    token = get_token()

    # Perform search
    data = search_wiki(args.query, token, args.limit)

    # Extract results
    results = data.get('results', [])
    total_size = data.get('totalSize', 0)

    if results:
        format_as_markdown(results)
        if total_size > len(results):
            print(f"\n_Showing {len(results)} of {total_size} results_")
    else:
        print("No results found.")


if __name__ == '__main__':
    main()

File: skills/supermap-wiki/scripts/search_wiki.js
#!/usr/bin/env node
/**
 * Supermap Wiki Search Script (Node.js Version)
 *
 * Searches the Supermap Confluence wiki and returns results as a markdown table.
 * Uses the SUPERMAP_WIKI_TOKEN environment variable for authentication.
 */

const https = require('https');

const WIKI_BASE_URL = 'wiki.ispeco.com';
const API_BASE_URL = '/rest/api/search';

function getToken() {
    const token = process.env.SUPERMAP_WIKI_TOKEN;
    if (!token) {
        console.error('Error: SUPERMAP_WIKI_TOKEN environment variable is not set.');
        console.error('Please set it using:');
        console.error('  Linux/macOS: export SUPERMAP_WIKI_TOKEN=\'your-token\'');
        console.error('  Windows (cmd): set SUPERMAP_WIKI_TOKEN=your-token');
        console.error('  Windows (PowerShell): $env:SUPERMAP_WIKI_TOKEN=\'your-token\'');
        process.exit(1);
    }
    return token;
}

function buildSearchUrl(query, limit = 20) {
    // Build CQL query
    const cql = `siteSearch ~ "${query}" AND type in ("space","user","com.atlassian.confluence.extra.team-calendars:calendar-content-type","attachment","page","com.atlassian.confluence.extra.team-calendars:space-calendars-view-content-type","blogpost")`;

    const params = new URLSearchParams({
        cql: cql,
        start: '0',
        limit: limit.toString(),
        excerpt: 'highlight',
        expand: 'space.icon',
        includeArchivedSpaces: 'false'
    });

    return `${API_BASE_URL}?${params.toString()}`;
}

function makeRequest(url, token) {
    return new Promise((resolve, reject) => {
        const options = {
            hostname: WIKI_BASE_URL,
            path: url,
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${token}`,
                'Accept': 'application/json',
                'User-Agent': 'Supermap-Wiki-Search/1.0'
            },
            timeout: 30000,
            rejectUnauthorized: false
        };

        const req = https.request(options, (res) => {
            let data = '';

            res.on('data', (chunk) => {
                data += chunk;
            });

            res.on('end', () => {
                if (res.statusCode >= 200 && res.statusCode < 300) {
                    try {
                        const jsonData = JSON.parse(data);
                        resolve(jsonData);
                    } catch (e) {
                        reject(new Error(`Failed to parse response: ${e.message}`));
                    }
                } else if (res.statusCode === 401) {
                    reject(new Error('Authentication failed. Please check your SUPERMAP_WIKI_TOKEN.'));
                } else if (res.statusCode === 403) {
                    reject(new Error('Access forbidden. You may not have permission to search the wiki.'));
                } else {
                    reject(new Error(`HTTP ${res.statusCode}`));
                }
            });
        });

        req.on('error', (err) => {
            reject(new Error(`Network error: ${err.message}`));
        });

        req.on('timeout', () => {
            req.destroy();
            reject(new Error('Request timeout'));
        });

        req.end();
    });
}

async function searchWiki(query, token, limit) {
    const url = buildSearchUrl(query, limit);

    try {
        const response = await makeRequest(url, token);
        return response;
    } catch (error) {
        console.error(`Error: ${error.message}`);
        process.exit(1);
    }
}

function cleanExcerpt(excerpt) {
    if (!excerpt) return '';

    // Remove HTML highlight tags and convert to markdown bold
    let cleaned = excerpt
        .replace(/<span class="search-highlight">/g, '**')
        .replace(/<\/span>/g, '**')
        .replace(/\n/g, ' ')
        .trim();

    // Truncate long excerpts
    if (cleaned.length > 150) {
        cleaned = cleaned.substring(0, 147) + '...';
    }

    return cleaned;
}

function escapePipe(str) {
    return str ? str.replace(/\|/g, '\\|') : '';
}

function formatAsMarkdown(results) {
    if (!results || !results.results || results.results.length === 0) {
        console.log('No results found.');
        return;
    }

    // Table header
    console.log('| Title | Space | Excerpt |');
    console.log('|-------|-------|---------|');

    // Table rows
    for (const result of results.results) {
        const content = result.content || {};
        const title = content.title || 'N/A';

        // Get space name
        const container = result.resultGlobalContainer || {};
        const space = container.title || 'N/A';

        // Get excerpt and clean it up
        const excerpt = cleanExcerpt(result.excerpt);

        // Build link
        const webuiLink = content._links?.webui || '';
        const titleWithLink = webuiLink
            ? `[${title}](https://${WIKI_BASE_URL}${webuiLink})`
            : title;

        // Escape pipe characters
        const escapedTitle = escapePipe(titleWithLink);
        const escapedSpace = escapePipe(space);
        const escapedExcerpt = escapePipe(excerpt);

        console.log(`| ${escapedTitle} | ${escapedSpace} | ${escapedExcerpt} |`);
    }

    const totalSize = results.totalSize || 0;
    const shownResults = results.results.length;
    if (totalSize > shownResults) {
        console.log(`\n_Showing ${shownResults} of ${totalSize} results_`);
    }
}

function printHelp() {
    console.log(`Supermap Wiki Search (Node.js Version)

Usage: node search_wiki.js <search-query> [options]

Description:
    Search Supermap wiki and display results as a markdown table.

Environment Variables:
    SUPERMAP_WIKI_TOKEN  - Required. Your wiki API token for authentication.

Options:
    -l, --limit <number>  Maximum number of results (default: 20)
    -h, --help           Show this help message

Examples:
    node search_wiki.js "project documentation"
    node search_wiki.js deployment -l 50
    node search_wiki.js --help

Output:
    Results are displayed as a markdown table with titles, spaces, and excerpts.
`);
}

function parseArgs(args) {
    const result = {
        query: '',
        limit: 20,
        help: false
    };

    for (let i = 0; i < args.length; i++) {
        const arg = args[i];

        if (arg === '-h' || arg === '--help') {
            result.help = true;
        } else if ((arg === '-l' || arg === '--limit') && i + 1 < args.length) {
            const limit = parseInt(args[i + 1], 10);
            if (!isNaN(limit) && limit > 0) {
                result.limit = limit;
            }
            i++;
        } else if (!result.query) {
            result.query = arg;
        } else {
            result.query += ' ' + arg;
        }
    }

    return result;
}

async function main() {
    const args = parseArgs(process.argv.slice(2));

    if (args.help || !args.query) {
        printHelp();
        process.exit(0);
    }

    const token = getToken();

    const data = await searchWiki(args.query, token, args.limit);
    formatAsMarkdown(data);
}

main().catch(err => {
    console.error(`Unexpected error: ${err.message}`);
    process.exit(1);
});
